frame/ranked-collective/src/lib.rs

```diff
@@ -0,0 +1,540 @@
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 2 votes;
+/// - ...with an excess rank of 3 gets 3 votes;
+/// - ...with an excess rank of 4 gets 4 votes.
+pub struct Linear;
+impl Convert<Rank, Votes> for Linear {
+	fn convert(r: Rank) -> Votes {
+		(r + 1) as Votes
+	}
+}
+
+/// Vote-weight scheme where all voters get one vote plus additional votes for every excess rank
+/// they have incrementing by one vote for each excess rank. I.e.:
+///
+/// - Each member with no excess rank gets 1 vote;
+/// - ...with an excess rank of 1 gets 2 votes;
+/// - ...with an excess rank of 2 gets 3 votes;
+/// - ...with an excess rank of 3 gets 6 votes;
+/// - ...with an excess rank of 4 gets 10 votes.
+pub struct Geometric;
+impl Convert<Rank, Votes> for Geometric {
+	fn convert(r: Rank) -> Votes {
+		let v = (r + 1) as Votes;
+		v * (v + 1) / 2
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The outer event type.
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The origin required to add, promote or remove a member.
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The polling system used for our voting.
+		type Polls: Polling<
+			TallyOf<Self, I>,
+			Votes = Votes,
+			Class = Rank,
+			Moment = Self::BlockNumber,
+		>;
+
+		/// Convert a rank_delta into a number of votes the rank gets.
+		///
+		/// Rank_delta is defined as the number of ranks above the minimum required to take part
+		/// in the poll.
+		type VoteWeight: Convert<Rank, Votes>;
+	}
+
+	/// The number of members in the collective who have at least the rank according to the index
+	/// of the vec.
+	#[pallet::storage]
+	pub type MemberCount<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Twox64Concat, Rank, MemberIndex, ValueQuery>;
+
+	/// The current members of the collective.
+	#[pallet::storage]
+	pub type Members<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Twox64Concat, T::AccountId, MemberRecord>;
+
+	/// The index of each ranks's member into the group of members who have at least that rank.
+	#[pallet::storage]
+	pub type IdToIndex<T: Config<I>, I: 'static = ()> =
+		StorageDoubleMap<_, Twox64Concat, Rank, Twox64Concat, T::AccountId, MemberIndex>;
+
+	/// The members in the collective by index. All indices in the range `0..MemberCount` will
+	/// return `Some`, however a member's index is not guaranteed to remain unchanged over time.
+	#[pallet::storage]
+	pub type IndexToId<T: Config<I>, I: 'static = ()> =
```

Comment:
I keep having a feeling that keeping both `IdToIndex` and `IndexToId` is a bit not ideal, as it seems to store _almost_ the same data, and yet we're spending some extra storage space for an extra double map, but I don't really seem to see a better way other than to sacrifice storage space for computational time, which doesn't exactly sound attractive either.